libs/fairness/src/lib/util/calculateFairnessMetric.ts

```diff
@@ -0,0 +1,198 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { IMetricResponse } from "@responsible-ai/core-ui";
+import _ from "lodash";
+
+import { FairnessModes } from "./FairnessMetrics";
+
+export function calculateFairnessMetric(
+  value: IMetricResponse,
+  fairnessMethod: FairnessModes
+) {
+  let response;
+  let binBounds;
+  let lowerBounds;
+  let upperBounds;
+  let minLowerBound;
+  let maxLowerBound;
+  let minUpperBound;
+  let maxUpperBound;
+
+  const bins = value.bins
+    .slice()
+    .filter((x) => x !== undefined && !Number.isNaN(x) && !_.isArray(x[0])); // filters out confidence bounds
+
+  const min = _.min(bins);
+  const max = _.max(bins);
+
+  if (min === undefined || max === undefined) {
+    return { overall: Number.NaN };
+  }
+
+  // Use confidence bounds for each bin if they exist
+  if (value?.binBounds) {
+    binBounds = value.binBounds
+      .slice()
+      .filter((x) => x !== undefined && !Number.isNaN(x));
+
+    lowerBounds = value?.binBounds.map((binBound) => {
+      return binBound.lower;
+    });
+    upperBounds = value?.binBounds.map((binBound) => {
+      return binBound.upper;
+    });
+    minLowerBound = _.min(lowerBounds);
+    maxLowerBound = _.max(lowerBounds);
+    minUpperBound = _.min(upperBounds);
+    maxUpperBound = _.max(upperBounds);
+  }
+
+  if (fairnessMethod === FairnessModes.Min) {
+    response = { overall: min };
+    if (minLowerBound && minUpperBound) {
+      response = {
+        ...response,
+        bounds: {
+          lower: minLowerBound,
+          upper: minUpperBound
+        }
+      };
+    }
+    return response;
+  }
+
+  if (fairnessMethod === FairnessModes.Max) {
+    response = { overall: max };
+    if (maxLowerBound && maxUpperBound) {
+      response = {
+        ...response,
+        bounds: {
+          lower: maxLowerBound,
+          upper: maxUpperBound
+        }
+      };
+    }
+    return response;
+  }
+  // Note: The `bounds` of this Ratio metric looks at all
+  // disaggregated bounds, even if those disaggregated bounds are not
+  // the bounds of the nominal extremes that contribute to the
+  // `overall` property of the Ratio
+  if (fairnessMethod === FairnessModes.Ratio) {
+    if (binBounds && binBounds.length > 1) {
+      let lowerRatio = Infinity;
+      let upperRatio = Infinity;
+      for (let i = 0; i < binBounds.length - 1; i++) {
+        for (let j = i + 1; j < binBounds.length; j++) {
+          let lowerCandidate;
+          let upperCandidate;
+
+          // if there is overlap
+          if (
+            binBounds[i].upper > binBounds[j].lower &&
+            binBounds[j].upper > binBounds[i].lower
+          ) {
+            const lowerCandidate1 = binBounds[i].lower / binBounds[j].upper;
```

Comment:
Might want an extra comment or two to outline what's going on through these sections (and through the difference calculations as well). Not too much, just something to make clear what's happening at each stage (e.g. go through all pairs, calculating the max and min ratios, noting the special case for overlap).

It's obvious now, but come back in three months, and a little guidance as to what's going on will be much appreciated.